include/quicr/moq_message_types.h

```diff
@@ -0,0 +1,286 @@
+  constexpr uint8_t MESSAGE_TYPE_SUBSCRIBE_ERROR          = 0x5;
+  constexpr uint8_t MESSAGE_TYPE_ANNOUNCE                 = 0x6;
+  constexpr uint8_t MESSAGE_TYPE_ANNOUNCE_OK              = 0x7;
+  constexpr uint8_t MESSAGE_TYPE_ANNOUNCE_ERROR           = 0x8;
+  constexpr uint8_t MESSAGE_TYPE_UNANNOUNCE               = 0x9;
+  constexpr uint8_t MESSAGE_TYPE_UNSUBSCRIBE              = 0xA;
+  constexpr uint8_t MESSAGE_TYPE_SUBSCRIBE_DONE           = 0xB;
+  constexpr uint8_t MESSAGE_TYPE_ANNOUNCE_CANCEL          = 0xC;
+  constexpr uint8_t MESSAGE_TYPE_TRACK_STATUS_REQUEST     = 0xD;
+  constexpr uint8_t MESSAGE_TYPE_TRACK_STATUS             = 0xE;
+  constexpr uint8_t MESSAGE_TYPE_GOAWAY                   = 0x10;
+  constexpr uint8_t MESSAGE_TYPE_CLIENT_SETUP             = 0x40;
+  constexpr uint8_t MESSAGE_TYPE_SERVER_SETUP             = 0x41;
+  constexpr uint8_t MESSAGE_TYPE_STREAM_HEADER_TRACK      = 0x50;
+  constexpr uint8_t MESSAGE_TYPE_STREAM_HEADER_GROUP      = 0x51;
+
+  // TODO (Suhas): rename it to StreamMapping
+  enum ForwardingPreference : uint8_t {
+    StreamPerGroup = 0,
+    StreamPerObject,
+    StreamPerPriority,
+    StreamPerTrack,
+    Datagram
+  };
+
+  //
+  // Parameters
+  //
+
+  enum struct ParameterType : uint8_t {
+    Role = 0x0,
+    Path = 0x1,
+    AuthorizationInfo = 0x2, // version specific, unused
+  };
+
+  struct MoqParameter {
+    ParamType param_type;
+    uintVar_t param_length;
+    bytes param_value;
+  };
+
+  //
+  // Setup
+  //
+
+  struct MoqClientSetup {
+    std::vector<Version> supported_versions;
+    MoqParameter role_parameter;
+    MoqParameter path_parameter;
+  };
+
+  struct MoqServerSetup {
+    Version supported_version;
+    MoqParameter role_parameter;
+    MoqParameter path_parameter;
+  };
+
+  MessageBuffer& operator<<(MessageBuffer &buffer, const MoqParameter &msg);
+  MessageBuffer& operator>>(MessageBuffer &buffer, MoqParameter &msg);
+  MessageBuffer& operator<<(MessageBuffer &buffer, const MoqClientSetup &msg);
+  MessageBuffer& operator>>(MessageBuffer &buffer, MoqClientSetup &msg);
+  MessageBuffer& operator<<(MessageBuffer &buffer, const MoqServerSetup &msg);
+  MessageBuffer& operator>>(MessageBuffer &buffer, MoqServerSetup &msg);
+
+    //
+  // Subscribe
+  //
+  enum struct LocationMode: uint8_t {
+      None = 0x0,
+      Absolute,
+      RelativePrevious,
+      RelativeNext
+  };
+
+  struct Location {
+    LocationMode mode;
+    std::optional<uintVar_t> value;
+  };
+
+  struct MoqSubscribe {
+    SubscribeId subscribe_id;
+    TrackAlias track_alias;
+    TrackNamespace track_namespace;
+    TrackName track_name;
+    Location start_group;
+    Location start_object;
+    Location end_group;
+    Location end_object;
+    std::vector<MoqParameter> track_params;
+ };
+
+  struct MoqSubscribeOk {
+    SubscribeId subscribe_id;
+    uintVar_t expires;
+    bool content_exists;
+    GroupId largest_group;
```

Comment:
largest group and object are optional based on `content_exists` right?

Reply:
that's right.